### DISCUSSION TITLE
How to properly type Generic Components with useController

### DISCUSSION THREAD

**Discussion by alveshelio** [react-hook-form/react-hook-form#8549] [Q&A]
Hi guys,

I'm trying to create a <ControlMultiSelect /> component with useController and downshift. I've created a codesandbox: https://codesandbox.io/s/usecontroller-chakra-ui-downshift-eb0bh8?file=/src/App.tsx and the component is working more or less, I still need to make sure that I filter selected components but this isn't the issue. Right now the component is throwing some TypeScript errors and I'm not quite sure how to solve these.

This component is generic, it's supposed to accept any kind of object that extends Items ({value:  string; label: string; }) 

But I'm not sure if I'm typing the component properly, I'm doing this
```ts
type WithCreationProps =
  | { withCreation: true; optionType: string }
  | { withCreation: false; optionType?: never };

interface Props<T> extends UseControllerProps<T>, UseMultipleSelectionProps<T> {
  items: T[] | GroupedTechnologyOption<T>[];
  placeholder: string;
  onSelectItems: (selectedItems?: T[]) => void;
  label?: string;
  helperText?: string;
  inputProps?: InputProps;
  containerStyles?: StackProps;
  listStyles?: ListProps;
  tagStyles?: TagProps;
  iconButtonStyles?: IconProps;
  initialSelectedItems?: T[];
  disabled?: boolean;
}

type MultiSelectProps<T> = Props<T> & WithCreationProps;

export const ControlMultiSelect = <T extends FieldValues>(props: MultiSelectProps<T>
```
I'm getting this error in items:
**(parameter) items: T[] | GroupedTechnologyOption<T>[]
Argument of type 'T[] | GroupedTechnologyOption<T>[]' is not assignable to parameter of type 'GroupedTechnologyOption<Option>[] | Option[]'.**

So if instead of doing `<T extends FieldValues>` I do `<T extends Option>` then I don't get errors, however when I use the component, name and control start giving me errors:
((Type '"countries"' is not assignable to type '"label" | "value"'.
Type 'Control<FormData, any>' is not assignable to type 'Control<Option, any>'.
The types of '_subjects.state' are incompatible between these types. **

I'm wondering how should a generic component be typed, in the examples in react-hook-form.com they have this `UseControllerProps<FormValues>`, however, this is for a very specific case, this field can only be used in this type. What I want is for the type to be inferred based on the types passed to useForm

**Comment by bluebill1049** [comment]
Building generic components with TS is always a bit of a challenge, especially with string template literal, hope this example helps a bit: 

https://codesandbox.io/s/vigilant-tu-c5stxd?file=/src/App.tsx

**Reply by alveshelio** [reply]
Hi @bluebill1049,

Thank you very much for the reply. I've made some changes made based on the codesandbox you've provided and I've provided a second generic type, now I'm doing
```ts
export const ControlMultiSelect = <T extends FieldValues, V extends Option>({
  // controller props
  name,
  control,
  // component props
  items,
  inputProps,
  iconButtonStyles,
  containerStyles,
  tagStyles,
  listStyles,
  initialSelectedItems = [],
  optionType,
  disabled,
  withCreation,
  label,
  helperText,
  ...downshiftProps
}: MultiSelectProps<T, V>): ReactElement => {
...
}
```
I'm almost there but I'm still getting an error on the controller.
**Type 'Control<FormData, any>' is not assignable to type 'Control<FieldValues, any>'.
  The types of '_options.resolver' are incompatible between these types.
    Type 'Resolver<FormData, any> | undefined' is not assignable to type 'Resolver<FieldValues, any> | undefined'.
      Type 'Resolver<FormData, any>' is not assignable to type 'Resolver<FieldValues, any>'.ts(2322)**
      
Not sure why this is still complaining though :(

Any ideas? :)
